lib/routing.js

```diff
@@ -29,13 +29,33 @@ const reverseMappings = mappings.map(field =>
     .join(':')
 )
 
+const serializeState = state => {
+  const stateString = encodeURIComponent(JSON.stringify(state))
+
+  return typeof window !== 'undefined'
+    ? btoa(stateString)
+    : Buffer.from(stateString).toString('base64')
+}
+
+const deserializeState = serializedState => {
+  let stateString
+  if (typeof window !== 'undefined') {
+    stateString = atob(serializedState)
+  } else {
+    stateString = Buffer.from(serializedState, 'base64').toString()
+  }
+
+  return JSON.parse(decodeURIComponent(stateString))
+}
+
 const keysToQuery = keys =>
   `?${Object.keys(keys)
     .map(key => `${key}=${keys[key]}`)
     .join('&')}`
 
-export const getQueryStringState = query => mapper.map(mappings, query)
+export const getQueryStringState = query =>
+  query.state ? deserializeState(query.state) : mapper.map(mappings, query)
 export const updateQueryString = state =>
   history.replace({
-    search: encodeURI(keysToQuery(mapper.map(reverseMappings, state)))
+    search: encodeURI(keysToQuery({ state: serializeState(state) }))
```

Comment:
@briandennis doesn't this remove the human-readable query params from the output? We want that so people can understand how to use it as an API without reading the code.

Reply:
Yes, that was for testing — should be fixed. Maybe worth adding an option to get the state string as it 1) currently supports options that normal params don't and 2) could be useful for debugging?

Reply:
Yes i thought you were just going to 1. First check if `state` is in the query params, if so decode that and 2. Fallback to the rest of the query params if state is not there.

Reply:
Yeah, that was the plan 👍 Just thinking now being able to gen a state dump might be useful?